diff --git a/gzip.c b/gzip.c
index 5d14022..e63cb44 100644
--- a/gzip.c
+++ b/gzip.c
@@ -74,6 +74,7 @@ static char const *const license_msg[] = {
 #include "ignore-value.h"
 #include "stat-time.h"
 #include "version.h"
+#include "yesno.h"
 
                 /* configuration */
 
diff --git a/gzip.h b/gzip.h
index b88cefb..c696914 100644
--- a/gzip.h
+++ b/gzip.h
@@ -325,5 +325,3 @@ extern void fprint_off    (FILE *, off_t, int);
         /* in inflate.c */
 extern int inflate (void);
 
-        /* in yesno.c */
-extern int yesno (void);
