diff --git a/lib/utimens.c b/lib/utimens.c
index 3e92c5c..f0b5a9a 100644
--- a/lib/utimens.c
+++ b/lib/utimens.c
@@ -27,6 +27,7 @@
 
 #include <errno.h>
 #include <fcntl.h>
+#include <sys/stat.h>
 #include <sys/time.h>
 #include <unistd.h>
 
@@ -78,6 +79,8 @@ int
 futimens_oi (int fd ATTRIBUTE_UNUSED,
 	  char const *file, struct timespec const timespec[2])
 {
+  int result;
+
   /* Some Linux-based NFS clients are buggy, and mishandle time stamps
      of files in NFS file systems in some cases.  We have no
      configure-time test for this, but please see
@@ -95,6 +99,24 @@ futimens_oi (int fd ATTRIBUTE_UNUSED,
     fsync (fd);
 #endif
 
+ /* attempt to use utimensat/futimens for nanosecond resolution.
+  * fall back to the old method if they don't succeed */
+
+ if (fd < 0)
+   {
+     result = utimensat (AT_FDCWD, file, timespec, 0);
+     if (result == 0)
+       return result;
+   }
+
+ if (0 <= fd)
+   {
+     result = futimens (fd, timespec);
+     if (result == 0)
+       return result;
+   }
+
+
   /* There's currently no interface to set file timestamps with
      nanosecond resolution, so do the best we can, discarding any
      fractional part of the timestamp.  */
